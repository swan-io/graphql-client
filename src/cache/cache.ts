import { Option, Result } from "@swan-io/boxed";
import { P, match } from "ts-pattern";
import {
  DEEP_MERGE_DELETE,
  containsAll,
  deepMerge,
  serializeVariables,
} from "../utils";
import {
  DocumentNode,
  OperationDefinitionNode,
  OperationTypeNode,
} from "@0no-co/graphql.web";

type CacheEntry = {
  requestedKeys: Set<symbol>;
  value: unknown;
};

export const getCacheKeyFromJson = (json: unknown): Option<symbol> => {
  return match(json)
    .with(
      { __typename: P.select(P.union("Query", "Mutation", "Subscription")) },
      (name) => Option.Some(Symbol.for(name))
    )
    .with(
      { __typename: P.select("name", P.string), id: P.select("id", P.string) },
      ({ name, id }) => Option.Some(Symbol.for(`${name}<${id}>`))
    )
    .otherwise(() => Option.None());
};

export const getCacheKeyFromOperationNode = (
  operationNode: OperationDefinitionNode
): Option<symbol> => {
  return match(operationNode.operation)
    .with(OperationTypeNode.QUERY, () => Option.Some(Symbol.for("Query")))
    .with(OperationTypeNode.SUBSCRIPTION, () =>
      Option.Some(Symbol.for("Subscription"))
    )
    .otherwise(() => Option.None());
};

const mergeCacheEntries = (a: CacheEntry, b: CacheEntry): CacheEntry => {
  return {
    requestedKeys: new Set([
      ...a.requestedKeys.values(),
      ...b.requestedKeys.values(),
    ]),
    value: deepMerge(a.value, b.value),
  };
};

export class ClientCache {
  cache = new Map<symbol, CacheEntry>();
  operationCache = new Map<
    DocumentNode,
    Map<string, Option<Result<unknown, unknown>>>
  >();

  dump() {
    return this.cache;
  }

  getOperationFromCache(
    documentNode: DocumentNode,
    variables: Record<string, any>
  ) {
    const serializedVariables = serializeVariables(variables);
    return Option.fromNullable(this.operationCache.get(documentNode))
      .flatMap((cache) => Option.fromNullable(cache.get(serializedVariables)))
      .flatMap((value) => value);
  }

  setOperationInCache(
    documentNode: DocumentNode,
    variables: Record<string, any>,
    data: Result<unknown, unknown>
  ) {
    const serializedVariables = serializeVariables(variables);
    const documentCache = Option.fromNullable(
      this.operationCache.get(documentNode)
    ).getWithDefault(new Map());
    documentCache.set(serializedVariables, Option.Some(data));
    this.operationCache.set(documentNode, documentCache);
  }

  getFromCache(cacheKey: symbol, requestedKeys: Set<symbol>) {
    return this.get(cacheKey).flatMap((entry) => {
      if (containsAll(entry.requestedKeys, requestedKeys)) {
        return Option.Some(entry.value);
      } else {
        return Option.None();
      }
    });
  }

  get(cacheKey: symbol): Option<CacheEntry> {
    if (this.cache.has(cacheKey)) {
      return Option.Some(this.cache.get(cacheKey) as CacheEntry);
    } else {
      return Option.None();
    }
  }

  getOrDefault(cacheKey: symbol): CacheEntry {
    if (this.cache.has(cacheKey)) {
      return this.cache.get(cacheKey) as CacheEntry;
    } else {
      return {
        requestedKeys: new Set(),
        value: {},
      };
    }
  }

  set(cacheKey: symbol, entry: CacheEntry) {
    this.cache.set(cacheKey, entry);
  }

  cacheIfEligible<T extends unknown>(
    value: T,
    requestedKeys: Set<symbol>
  ): symbol | T {
    return match(getCacheKeyFromJson(value))
      .with(Option.P.Some(P.select()), (cacheKey) => {
        const existingEntry = this.getOrDefault(cacheKey);
        this.cache.set(
          cacheKey,
          mergeCacheEntries(existingEntry, {
            requestedKeys,
            value,
          })
        );
        return cacheKey;
      })
      .otherwise(() => value);
  }

  updateFieldInClosestCachedAncestor({
    originalFieldName,
    fieldNameWithArguments,
    value,
    path,
    ancestors,
  }: {
    originalFieldName: string;
    fieldNameWithArguments: symbol | string;
    value: unknown;
    path: PropertyKey[];
    ancestors: any[];
  }) {
    const ancestorsCopy = ancestors.concat();
    const pathCopy = path.concat();
    const writePath: PropertyKey[] = [];

    let ancestor;

    while ((ancestor = ancestorsCopy.pop())) {
      const maybeCacheKey = getCacheKeyFromJson(ancestor);
      if (maybeCacheKey.isSome()) {
        const cacheKey = maybeCacheKey.get();
        const existingEntry = this.getOrDefault(cacheKey);

        const deepUpdate = writePath.reduce<unknown>(
          (acc, key) => {
            return {
              [key]: acc,
            };
          },
          // remote original field
          {
            [originalFieldName]: DEEP_MERGE_DELETE,
            [fieldNameWithArguments]: value,
          }
        );

        this.set(
          cacheKey,
          mergeCacheEntries(existingEntry, {
            requestedKeys: new Set(),
            value: deepUpdate,
          })
        );

        // When the cached ancestor is found, remove
        break;
      }

      writePath.push(pathCopy.pop() as PropertyKey);
    }
  }
}
