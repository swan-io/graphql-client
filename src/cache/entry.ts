import { CONNECTION_REF, REQUESTED_KEYS } from "../utils";

export type CacheEntry = Record<symbol, unknown> & {
  [REQUESTED_KEYS]: Set<symbol>;
  [CONNECTION_REF]?: number;
};

export const createEmptyCacheEntry = (): CacheEntry => ({
  [REQUESTED_KEYS]: new Set<symbol>(),
});
