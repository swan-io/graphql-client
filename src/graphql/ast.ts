import {
  ASTNode,
  DirectiveNode,
  DocumentNode,
  FieldNode,
  FragmentDefinitionNode,
  InlineFragmentNode,
  Kind,
  OperationDefinitionNode,
  OperationTypeNode,
  SelectionNode,
  SelectionSetNode,
  ValueNode,
  visit,
} from "@0no-co/graphql.web";
import { Array, Option } from "@swan-io/boxed";

/**
 * Returns a Set<string> with all keys selected within the direct selection sets
 * of a given `FieldNode` or `OperationDefinitionNode`.
 *
 * { user { id, firstName, lastName } }
 * => Set{"id", "firstName", "lastName"}
 *
 * @param fieldNode FieldNode | OperationDefinitionNode
 * @returns selectedKeys Set<string>
 */
export const getSelectedKeys = (
  fieldNode: FieldNode | OperationDefinitionNode,
  variables: Record<string, unknown>,
): Set<symbol> => {
  const selectedKeys = new Set<symbol>();

  const traverse = (selections: SelectionSetNode) => {
    // We only need to care about FieldNode & InlineFragment node
    // as we inline all fragments in the query
    selections.selections.forEach((selection) => {
      if (selection.kind === Kind.FIELD) {
        const fieldNameWithArguments = getFieldNameWithArguments(
          selection,
          variables,
        );
        selectedKeys.add(fieldNameWithArguments);
      } else if (selection.kind === Kind.INLINE_FRAGMENT) {
        traverse(selection.selectionSet);
      }
    });
  };

  if (fieldNode.selectionSet) {
    traverse(fieldNode.selectionSet);
  }

  return selectedKeys;
};

/**
 * Serializes the field name and arguments as a symbol.
 *
 * { user {id} }
 * => Symbol(`user`)
 *
 * { user(id: "1") {id} }
 * => Symbol(`user({"id":"1"})`)
 *
 * { user(id: $id) {id} } with variables `{"id": "2"}`
 * => Symbol(`user({"id":"2"})`)
 *
 * @param fieldNode
 * @param variables The variables of the GraphQL operation
 * @returns symbol
 */
export const getFieldNameWithArguments = (
  fieldNode: FieldNode,
  variables: Record<string, unknown>,
): symbol => {
  const fieldName = getFieldName(fieldNode);
  const args = extractArguments(fieldNode, variables);
  if (Object.keys(args).length === 0) {
    return Symbol.for(fieldName);
  }
  return Symbol.for(`${fieldName}(${JSON.stringify(args)})`);
};

/**
 * Returns a record representation of the arguments passed to a given field
 *
 * @param fieldNode
 * @param variables
 * @returns Record<string, any>
 */
export const extractArguments = (
  fieldNode: FieldNode,
  variables: Record<string, unknown>,
): Record<string, unknown> => {
  const args = fieldNode.arguments ?? [];
  return Object.fromEntries(
    args.map(({ name: { value: name }, value }) => [
      name,
      extractValue(value, variables),
    ]),
  );
};

/**
 * Resolves and serializes a GraphQL value
 *
 * @param valueNode: ValueNode
 * @param variables: Record<string, any>
 * @returns Record<string, any>
 */
const extractValue = (
  valueNode: ValueNode,
  variables: Record<string, unknown>,
): unknown => {
  switch (valueNode.kind) {
    case Kind.NULL:
      return null;
    case Kind.INT:
    case Kind.FLOAT:
    case Kind.STRING:
    case Kind.BOOLEAN:
    case Kind.ENUM:
      return valueNode.value;
    case Kind.LIST:
      return valueNode.values.map((value) => extractValue(value, variables));
    case Kind.OBJECT:
      return Object.fromEntries(
        valueNode.fields.map(({ name: { value: name }, value }) => [
          name,
          extractValue(value, variables),
        ]),
      );
    case Kind.VARIABLE:
      return variables[valueNode.name.value];
    default:
      return null;
  }
};

/**
 * Gets the field name in the response payload from its AST definition
 *
 * @param fieldNode
 * @returns field name
 */
export const getFieldName = (fieldNode: FieldNode) => {
  return fieldNode.alias ? fieldNode.alias.value : fieldNode.name.value;
};

/**
 * Simplifies the query for internal processing by inlining all fragments.
 *
 * @param documentNode
 * @returns documentNode
 */
export const inlineFragments = (documentNode: DocumentNode): DocumentNode => {
  const fragmentMap: { [fragmentName: string]: FragmentDefinitionNode } = {};

  // Populate the fragment map
  visit(documentNode, {
    [Kind.FRAGMENT_DEFINITION](node: FragmentDefinitionNode) {
      fragmentMap[node.name.value] = node;
    },
  });

  const inline = (node: ASTNode): unknown => {
    if (node.kind === Kind.FRAGMENT_SPREAD) {
      const fragmentName = node.name.value;
      const fragmentNode = fragmentMap[fragmentName];
      if (!fragmentNode) {
        throw new Error(`Fragment "${fragmentName}" is not defined.`);
      }
      const nextNode: InlineFragmentNode = {
        kind: Kind.INLINE_FRAGMENT,
        typeCondition: fragmentNode.typeCondition,
        selectionSet: fragmentNode.selectionSet,
      };
      return nextNode;
    }

    if (node.kind === Kind.SELECTION_SET) {
      return {
        ...node,
        selections: node.selections.map((selection: SelectionNode) =>
          inline(selection),
        ),
      };
    }

    if ("selectionSet" in node && node.selectionSet != null) {
      return {
        ...node,
        selectionSet: inline(node.selectionSet),
      };
    }

    return node;
  };

  return visit(documentNode, {
    [Kind.FRAGMENT_DEFINITION]: () => null,
    enter: inline,
  });
};

const TYPENAME_NODE: FieldNode = {
  kind: Kind.FIELD,
  name: {
    kind: Kind.NAME,
    value: "__typename",
  },
};

/**
 * Adds `__typename` to all selection sets in the document
 *
 * @param documentNode
 * @returns documentNode
 */
export const addTypenames = (documentNode: DocumentNode): DocumentNode => {
  return visit(documentNode, {
    [Kind.SELECTION_SET]: (selectionSet): SelectionSetNode => {
      if (
        selectionSet.selections.find(
          (selection) =>
            selection.kind === Kind.FIELD &&
            selection.name.value === "__typename",
        )
      ) {
        return selectionSet;
      } else {
        return {
          ...selectionSet,
          selections: [TYPENAME_NODE, ...selectionSet.selections],
        };
      }
    },
  });
};

export const getExecutableOperationName = (document: DocumentNode) => {
  return Array.findMap(document.definitions, (definition) => {
    if (definition.kind === Kind.OPERATION_DEFINITION) {
      return Option.fromNullable(definition.name).map((name) => name.value);
    } else {
      return Option.None();
    }
  });
};

const getIdFieldNode = (selection: SelectionNode): Option<SelectionNode> => {
  switch (selection.kind) {
    case Kind.FIELD:
      return selection.name.value === "id"
        ? Option.Some(selection)
        : Option.None();
    case Kind.INLINE_FRAGMENT:
      return Array.findMap(selection.selectionSet.selections, getIdFieldNode);
    default:
      return Option.None();
  }
};

export const addIdIfPreviousSelected = (
  oldSelectionSet: SelectionSetNode,
  newSelectionSet: SelectionSetNode,
): SelectionSetNode => {
  const idSelection = Array.findMap(oldSelectionSet.selections, getIdFieldNode);
  const idSelectionInNew = Array.findMap(
    newSelectionSet.selections,
    getIdFieldNode,
  );

  if (idSelectionInNew.isSome()) {
    return newSelectionSet;
  }

  return idSelection
    .map((selection) => ({
      ...newSelectionSet,
      selections: [
        selection,
        ...newSelectionSet.selections,
      ] as readonly SelectionNode[],
    }))
    .getOr(newSelectionSet);
};

export const isExcluded = (
  fieldNode: FieldNode,
  variables: Record<string, unknown>,
) => {
  if (!Array.isArray(fieldNode.directives)) {
    return false;
  }

  return fieldNode.directives.some(
    (directive: DirectiveNode) =>
      directive.name.value === "include" &&
      directive.arguments != null &&
      directive.arguments.some((arg) => {
        return (
          arg.name.value === "if" &&
          extractValue(arg.value, variables) === false
        );
      }),
  );
};

export const getCacheKeyFromOperationNode = (
  operationNode: OperationDefinitionNode,
): Option<symbol> => {
  switch (operationNode.operation) {
    case OperationTypeNode.QUERY:
      return Option.Some(Symbol.for("Query"));
    case OperationTypeNode.SUBSCRIPTION:
      return Option.Some(Symbol.for("Subscription"));
    default:
      return Option.None();
  }
};
