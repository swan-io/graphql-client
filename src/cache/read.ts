import { DocumentNode, SelectionSetNode, Kind } from "@0no-co/graphql.web";
import { ClientCache, getCacheKeyFromOperationNode } from "./cache";
import {
  getFieldName,
  getFieldNameWithArguments,
  getSelectedKeys,
} from "../graphql/ast";
import { match } from "ts-pattern";
import { Array, Option } from "@swan-io/boxed";
import { deepEqual, isRecord } from "../utils";

const getFromCacheOrReturnValue = (
  cache: ClientCache,
  valueOrKey: unknown,
  selectedKeys: Set<string>
): Option<unknown> =>
  typeof valueOrKey === "symbol"
    ? cache.getFromCache(valueOrKey, selectedKeys).flatMap(Option.fromNullable)
    : Option.Some(valueOrKey);

const STABILITY_CACHE = new WeakMap<DocumentNode, Map<string, unknown>>();

export const readOperationFromCache = (
  cache: ClientCache,
  document: DocumentNode,
  variables: Record<string, any>
) => {
  const traverse = (
    selections: SelectionSetNode,
    data: Record<PropertyKey, unknown>
  ): Option<any> => {
    return selections.selections.reduce((data, selection) => {
      return data.flatMap((data) =>
        match(selection)
          .with({ kind: Kind.FIELD }, (fieldNode) => {
            const originalFieldName = getFieldName(fieldNode);
            const fieldNameWithArguments = getFieldNameWithArguments(
              fieldNode,
              variables
            );
            if (data == undefined) {
              return Option.None();
            }
            const valueOrKeyFromCache = data[fieldNameWithArguments];

            if (valueOrKeyFromCache == undefined) {
              return Option.Some({
                ...data,
                [originalFieldName]: valueOrKeyFromCache,
              });
            }

            if (Array.isArray(valueOrKeyFromCache)) {
              const selectedKeys = getSelectedKeys(fieldNode);
              return Option.all(
                valueOrKeyFromCache.map((valueOrKey) => {
                  const value = getFromCacheOrReturnValue(
                    cache,
                    valueOrKey,
                    selectedKeys
                  );

                  return value.flatMap((value) => {
                    if (
                      isRecord(value) &&
                      fieldNode.selectionSet != undefined
                    ) {
                      return traverse(fieldNode.selectionSet, value);
                    } else {
                      return Option.Some(value);
                    }
                  });
                })
              ).map((result) => ({
                ...data,
                [originalFieldName]: result,
              }));
            } else {
              const selectedKeys = getSelectedKeys(fieldNode);

              const value = getFromCacheOrReturnValue(
                cache,
                valueOrKeyFromCache,
                selectedKeys
              );

              return value.flatMap((value) => {
                if (isRecord(value) && fieldNode.selectionSet != undefined) {
                  return traverse(
                    fieldNode.selectionSet,
                    value as Record<PropertyKey, unknown>
                  ).map((result) => ({
                    ...data,
                    [originalFieldName]: result,
                  }));
                } else {
                  return Option.Some({ ...data, [originalFieldName]: value });
                }
              });
            }
          })
          .with({ kind: Kind.INLINE_FRAGMENT }, (inlineFragmentNode) => {
            return traverse(inlineFragmentNode.selectionSet, data);
          })
          .with({ kind: Kind.FRAGMENT_SPREAD }, () => {
            return Option.None();
          })
          .exhaustive()
      );
    }, Option.Some(data));
  };

  return Array.findMap(document.definitions, (definition) =>
    definition.kind === Kind.OPERATION_DEFINITION
      ? Option.Some(definition)
      : Option.None()
  )
    .flatMap((operation) =>
      getCacheKeyFromOperationNode(operation).map((cacheKey) => ({
        operation,
        cacheKey,
      }))
    )
    .flatMap(({ operation, cacheKey }) => {
      return cache
        .getFromCache(cacheKey, getSelectedKeys(operation))
        .map((cache) => ({ cache, operation }));
    })
    .flatMap(({ operation, cache }) => {
      return traverse(
        operation.selectionSet,
        cache as Record<PropertyKey, unknown>
      );
    })
    .map((data) => JSON.parse(JSON.stringify(data)))
    .flatMap((value) => {
      // We use a trick to return stable values, the document holds a WeakMap
      // that for each key (serialized variables), stores the last returned result.
      // If the last value deeply equals the previous one, return the previous one
      const serializedVariables = JSON.stringify(variables);
      const previous = Option.fromNullable(STABILITY_CACHE.get(document))
        .flatMap((byVariable) =>
          Option.fromNullable(byVariable.get(serializedVariables))
        )
        .flatMap((value) => value as Option<unknown>);

      if (previous.isSome() && deepEqual(value, previous.get())) {
        return previous;
      } else {
        const valueToCache = Option.Some(value);
        const documentCache = STABILITY_CACHE.get(document) ?? new Map();
        documentCache.set(serializedVariables, valueToCache);
        STABILITY_CACHE.set(document, documentCache);
        return valueToCache;
      }
    });
};
