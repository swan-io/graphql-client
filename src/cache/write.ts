import { DocumentNode, SelectionSetNode, Kind } from "@0no-co/graphql.web";
import { ClientCache } from "./cache";
import {
  extractArguments,
  getFieldName,
  getFieldNameWithArguments,
  getSelectedKeys,
} from "../graphql/ast";
import { match } from "ts-pattern";
import { isRecord } from "../utils";

export const writeOperationToCache = (
  cache: ClientCache,
  document: DocumentNode,
  response: any,
  variables: Record<string, any>
) => {
  const traverse = (
    selections: SelectionSetNode,
    data: any[],
    path: PropertyKey[] = []
  ) => {
    selections.selections.forEach((selection) => {
      match(selection)
        .with({ kind: Kind.FIELD }, (fieldNode) => {
          const originalFieldName = getFieldName(fieldNode);
          const fieldNameWithArguments = getFieldNameWithArguments(
            fieldNode,
            variables
          );
          const fieldArguments = extractArguments(fieldNode, variables);

          const fieldValue = data.at(-1)[originalFieldName];
          const selectedKeys = getSelectedKeys(fieldNode, variables);

          if (fieldValue != undefined) {
            if (Array.isArray(fieldValue)) {
              fieldValue.forEach((item: any, index: number) => {
                const value = cache.cacheIfEligible(item, selectedKeys);

                const nextValue = Array(fieldValue.length);
                nextValue[index] = value;

                cache.updateFieldInClosestCachedAncestor({
                  originalFieldName,
                  fieldNameWithArguments,
                  value: nextValue,
                  path,
                  ancestors: data,
                  variables: fieldArguments,
                });

                if (isRecord(item) && !Array.isArray(item)) {
                  traverse(
                    fieldNode.selectionSet!,
                    [...data, fieldValue, item],
                    [...path, fieldNameWithArguments, index.toString()]
                  );
                }
              });
            } else {
              const value = cache.cacheIfEligible(fieldValue, selectedKeys);

              cache.updateFieldInClosestCachedAncestor({
                originalFieldName,
                fieldNameWithArguments,
                value,
                path,
                ancestors: data,
                variables: fieldArguments,
              });

              if (isRecord(fieldValue) && fieldNode.selectionSet != undefined) {
                traverse(
                  fieldNode.selectionSet,
                  [...data, fieldValue],
                  [...path, fieldNameWithArguments]
                );
              }
            }
          } else {
            cache.updateFieldInClosestCachedAncestor({
              originalFieldName,
              fieldNameWithArguments,
              value: fieldValue,
              path,
              ancestors: data,
              variables: fieldArguments,
            });
          }
        })
        .with({ kind: Kind.INLINE_FRAGMENT }, (inlineFragmentNode) => {
          traverse(inlineFragmentNode.selectionSet, data, path);
        })
        .with({ kind: Kind.FRAGMENT_SPREAD }, () => {
          // ignore, those are stripped
        })
        .exhaustive();
    });
  };

  document.definitions.forEach((definition) => {
    if (definition.kind === Kind.OPERATION_DEFINITION) {
      cache.cacheIfEligible(response, getSelectedKeys(definition, variables));
      traverse(definition.selectionSet, [response]);
    }
  });

  return cache;
};
