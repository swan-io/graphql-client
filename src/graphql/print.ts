import { ASTNode } from "@0no-co/graphql.web";

export const printString = (string: string) => {
  return JSON.stringify(string);
};

export const printBlockString = (string: string) => {
  return '"""' + string.replace(/"""/g, '\\"""') + '"""';
};

const hasItems = <T>(
  array: ReadonlyArray<T> | undefined | null,
): array is ReadonlyArray<T> => Boolean(array && array.length);

const MAX_LINE_LENGTH = 80;

const nodes: {
  [NodeT in ASTNode as NodeT["kind"]]?: (node: NodeT) => string;
} = {
  OperationDefinition(node) {
    if (
      node.operation === "query" &&
      !node.name &&
      !hasItems(node.variableDefinitions) &&
      !hasItems(node.directives)
    ) {
      return nodes.SelectionSet!(node.selectionSet);
    }
    let out: string = node.operation;
    if (node.name) out += " " + node.name.value;
    if (hasItems(node.variableDefinitions)) {
      if (!node.name) out += " ";
      out +=
        "(" +
        node.variableDefinitions.map(nodes.VariableDefinition!).join(", ") +
        ")";
    }
    if (hasItems(node.directives))
      out += " " + node.directives.map(nodes.Directive!).join(" ");
    return out + " " + nodes.SelectionSet!(node.selectionSet);
  },
  VariableDefinition(node) {
    let out = nodes.Variable!(node.variable) + ": " + print(node.type);
    if (node.defaultValue) out += " = " + print(node.defaultValue);
    if (hasItems(node.directives))
      out += " " + node.directives.map(nodes.Directive!).join(" ");
    return out;
  },
  Field(node) {
    let out = (node.alias ? node.alias.value + ": " : "") + node.name.value;
    if (hasItems(node.arguments)) {
      const args = node.arguments.map(nodes.Argument!);
      const argsLine = out + "(" + args.join(", ") + ")";
      out =
        argsLine.length > MAX_LINE_LENGTH
          ? out + "(" + args.join(" ") + ")"
          : argsLine;
    }
    if (hasItems(node.directives))
      out += " " + node.directives.map(nodes.Directive!).join(" ");
    return node.selectionSet
      ? out + " " + nodes.SelectionSet!(node.selectionSet)
      : out;
  },
  StringValue(node) {
    return node.block ? printBlockString(node.value) : printString(node.value);
  },
  BooleanValue(node) {
    return String(node.value);
  },
  NullValue() {
    return "null";
  },
  IntValue(node) {
    return node.value;
  },
  FloatValue(node) {
    return node.value;
  },
  EnumValue(node) {
    return node.value;
  },
  Name(node) {
    return node.value;
  },
  Variable(node) {
    return "$" + node.name.value;
  },
  ListValue(node) {
    return "[" + node.values.map(print).join(", ") + "]";
  },
  ObjectValue(node) {
    return "{" + node.fields.map(nodes.ObjectField!).join(", ") + "}";
  },
  ObjectField(node) {
    return node.name.value + ": " + print(node.value);
  },
  Document(node) {
    return hasItems(node.definitions)
      ? node.definitions.map(print).join(" ")
      : "";
  },
  SelectionSet(node) {
    return "{" + node.selections.map(print).join(" ") + "}";
  },
  Argument(node) {
    return node.name.value + ": " + print(node.value);
  },
  FragmentSpread(node) {
    let out = "..." + node.name.value;
    if (hasItems(node.directives))
      out += " " + node.directives.map(nodes.Directive!).join(" ");
    return out;
  },
  InlineFragment(node) {
    let out = "...";
    if (node.typeCondition) out += " on " + node.typeCondition.name.value;
    if (hasItems(node.directives))
      out += " " + node.directives.map(nodes.Directive!).join(" ");
    return out + " " + print(node.selectionSet);
  },
  FragmentDefinition(node) {
    let out = "fragment " + node.name.value;
    out += " on " + node.typeCondition.name.value;
    if (hasItems(node.directives))
      out += " " + node.directives.map(nodes.Directive!).join(" ");
    return out + " " + print(node.selectionSet);
  },
  Directive(node) {
    let out = "@" + node.name.value;
    if (hasItems(node.arguments))
      out += "(" + node.arguments.map(nodes.Argument!).join(", ") + ")";
    return out;
  },
  NamedType(node) {
    return node.name.value;
  },
  ListType(node) {
    return "[" + print(node.type) + "]";
  },
  NonNullType(node) {
    return print(node.type) + "!";
  },
};

export const print = (node: ASTNode): string => {
  // @ts-expect-error It's safe
  return typeof nodes[node.kind] == "function" ? nodes[node.kind](node) : "";
};
