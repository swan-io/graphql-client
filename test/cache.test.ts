import { Option, Result } from "@swan-io/boxed";
import { expect, test } from "vitest";
import { ClientCache } from "../src/cache/cache";
import { optimizeQuery, readOperationFromCache } from "../src/cache/read";
import { writeOperationToCache } from "../src/cache/write";
import { addTypenames, inlineFragments } from "../src/graphql/ast";
import { print } from "../src/graphql/print";
import {
  appQuery,
  appQueryWithExtraArrayInfo,
  bindAccountMembershipMutation,
  bindMembershipMutationRejectionResponse,
  bindMembershipMutationSuccessResponse,
  getAppQueryResponse,
  otherAppQuery,
} from "./data";

test("Write & read in cache", () => {
  const cache = new ClientCache();

  const preparedAppQuery = inlineFragments(addTypenames(appQuery));

  writeOperationToCache(
    cache,
    preparedAppQuery,
    getAppQueryResponse({
      user2LastName: "Last",
      user1IdentificationLevels: null,
    }),
    {
      id: "1",
    },
  );

  expect(cache.dump()).toMatchSnapshot();

  expect(
    readOperationFromCache(cache, preparedAppQuery, {
      id: "1",
    }),
  ).toMatchObject(
    Option.Some(
      Result.Ok(
        getAppQueryResponse({
          user2LastName: "Last",
          user1IdentificationLevels: null,
        }),
      ),
    ),
  );

  const preparedBindAccountMembershipMutation = inlineFragments(
    addTypenames(bindAccountMembershipMutation),
  );

  const preparedOtherAppQuery = inlineFragments(addTypenames(otherAppQuery));
  const preparedAppQueryWithExtraArrayInfo = inlineFragments(
    addTypenames(appQueryWithExtraArrayInfo),
  );

  writeOperationToCache(
    cache,
    preparedBindAccountMembershipMutation,
    bindMembershipMutationRejectionResponse,
    {
      id: "account-membership-2",
    },
  );

  expect(
    readOperationFromCache(cache, preparedAppQuery, {
      id: "1",
    }),
  ).toMatchObject(
    Option.Some(
      Result.Ok(
        getAppQueryResponse({
          user2LastName: "Last",
          user1IdentificationLevels: null,
        }),
      ),
    ),
  );

  writeOperationToCache(
    cache,
    preparedBindAccountMembershipMutation,
    bindMembershipMutationSuccessResponse,
    {
      id: "account-membership-2",
    },
  );

  expect(cache.dump()).toMatchSnapshot();

  expect(
    readOperationFromCache(cache, preparedAppQuery, {
      id: "1",
    }),
  ).toMatchObject(
    Option.Some(
      Result.Ok(
        getAppQueryResponse({
          user2LastName: "Acthernoene",
          user1IdentificationLevels: null,
        }),
      ),
    ),
  );

  writeOperationToCache(
    cache,
    preparedAppQuery,
    getAppQueryResponse({
      user2LastName: "Acthernoene",
      user1IdentificationLevels: {
        expert: true,
        PVID: true,
        QES: true,
      },
    }),
    {
      id: "1",
    },
  );

  expect(cache.dump()).toMatchSnapshot();

  expect(
    readOperationFromCache(cache, preparedAppQuery, {
      id: "1",
    }),
  ).toMatchObject(
    Option.Some(
      Result.Ok(
        getAppQueryResponse({
          user2LastName: "Acthernoene",
          user1IdentificationLevels: {
            expert: true,
            PVID: true,
            QES: true,
          },
        }),
      ),
    ),
  );

  const values = Option.all([
    readOperationFromCache(cache, preparedAppQuery, {
      id: "1",
    }),
    readOperationFromCache(cache, preparedAppQuery, {
      id: "1",
    }),
  ]);

  if (values.isSome()) {
    const [a, b] = values.get();
    expect(a).toBe(b);
  } else {
    expect(true).toBe(false);
  }

  expect(
    optimizeQuery(cache, preparedOtherAppQuery, { id: "1" })
      .map(print)
      .getWithDefault("no delta"),
  ).toMatchSnapshot();

  expect(
    optimizeQuery(cache, preparedAppQueryWithExtraArrayInfo, { id: "1" })
      .map(print)
      .getWithDefault("no delta"),
  ).toMatchSnapshot();
});
