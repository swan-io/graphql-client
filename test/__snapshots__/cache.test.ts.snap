// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Write & read in cache 1`] = `
Map {
  Symbol(Query) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(accountMembership({"id":"1"})),
      Symbol(accountMemberships({"first":"2"})),
      Symbol(supportingDocumentCollection({"id":"e8d38e87-9862-47ef-b749-212ed566b955"})),
    },
    "value": {
      Symbol(__typename): "Query",
      Symbol(accountMembership({"id":"1"})): Symbol(AccountMembership<account-membership-1>),
      Symbol(accountMemberships({"first":"2"})): {
        Symbol(__typename): "AccountMembershipConnection",
        Symbol(edges): [
          {
            Symbol(__typename): "AccountMembershipEdge",
            Symbol(node): Symbol(AccountMembership<account-membership-1>),
          },
          {
            Symbol(__typename): "AccountMembershipEdge",
            Symbol(node): Symbol(AccountMembership<account-membership-2>),
          },
        ],
      },
      Symbol(supportingDocumentCollection({"id":"e8d38e87-9862-47ef-b749-212ed566b955"})): Symbol(SupportingDocumentCollection<supporting-document-collection-1>),
    },
  },
  Symbol(AccountMembership<account-membership-1>) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(id),
      Symbol(user),
      Symbol(account),
      Symbol(membershipUser),
    },
    "value": {
      Symbol(__typename): "AccountMembership",
      Symbol(id): "account-membership-1",
      Symbol(user): Symbol(User<user-1>),
      Symbol(account): {
        Symbol(__typename): "Account",
        Symbol(name): "First",
      },
      Symbol(membershipUser): Symbol(User<user-1>),
    },
  },
  Symbol(User<user-1>) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(id),
      Symbol(firstName),
      Symbol(lastName),
      Symbol(identificationLevels),
    },
    "value": {
      Symbol(__typename): "User",
      Symbol(id): "user-1",
      Symbol(firstName): "Matthias",
      Symbol(lastName): "Le Brun",
      Symbol(identificationLevels): null,
    },
  },
  Symbol(AccountMembership<account-membership-2>) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(id),
      Symbol(account),
      Symbol(membershipUser),
    },
    "value": {
      Symbol(__typename): "AccountMembership",
      Symbol(id): "account-membership-2",
      Symbol(account): {
        Symbol(__typename): "Account",
        Symbol(name): "Second",
      },
      Symbol(membershipUser): Symbol(User<user-2>),
    },
  },
  Symbol(User<user-2>) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(id),
      Symbol(lastName),
    },
    "value": {
      Symbol(__typename): "User",
      Symbol(id): "user-2",
      Symbol(lastName): "Last",
    },
  },
  Symbol(SupportingDocumentCollection<supporting-document-collection-1>) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(supportingDocuments),
      Symbol(id),
    },
    "value": {
      Symbol(__typename): "SupportingDocumentCollection",
      Symbol(supportingDocuments): [
        Symbol(SupportingDocument<supporting-document-1>),
      ],
      Symbol(id): "supporting-document-collection-1",
    },
  },
  Symbol(SupportingDocument<supporting-document-1>) => {
    "requestedKeys": Set {
      Symbol(__typename),
      Symbol(id),
      Symbol(createdAt),
    },
    "value": {
      Symbol(__typename): "SupportingDocument",
      Symbol(id): "supporting-document-1",
      Symbol(createdAt): "2024-03-14T12:06:10.857Z",
    },
  },
}
`;
